python/ray/autoscaler/local/node_provider.py

```diff
@@ -19,58 +20,62 @@
+                            "Loaded cluster state: {}".format(workers))
+                for worker_ip in provider_config["worker_ips"]:
+                    if worker_ip not in workers:
+                        workers[worker_ip] = {
+                            "tags": {
+                                TAG_RAY_NODE_TYPE: "worker"
+                            },
+                            "state": "terminated",
+                        }
+                    else:
+                        assert workers[worker_ip]["tags"][
+                            TAG_RAY_NODE_TYPE] == "worker"
+                if provider_config["head_ip"] not in workers:
+                    workers[provider_config["head_ip"]] = {
                         "tags": {
-                            TAG_RAY_NODE_TYPE: "worker"
+                            TAG_RAY_NODE_TYPE: "head"
                         },
                         "state": "terminated",
                     }
                 else:
-                    assert workers[worker_ip]["tags"][
-                        TAG_RAY_NODE_TYPE] == "worker"
-            if provider_config["head_ip"] not in workers:
-                workers[provider_config["head_ip"]] = {
-                    "tags": {
-                        TAG_RAY_NODE_TYPE: "head"
-                    },
-                    "state": "terminated",
-                }
-            else:
-                assert workers[provider_config["head_ip"]]["tags"][
-                    TAG_RAY_NODE_TYPE] == "head"
-            assert len(workers) == len(provider_config["worker_ips"]) + 1
-            with open(self.save_path, "w") as f:
-                logger.info("ClusterState: "
-                            "Writing cluster state: {}".format(workers))
-                f.write(json.dumps(workers))
+                    assert workers[provider_config["head_ip"]]["tags"][
+                        TAG_RAY_NODE_TYPE] == "head"
+                assert len(workers) == len(provider_config["worker_ips"]) + 1
+                with open(self.save_path, "w") as f:
+                    logger.info("ClusterState: "
+                                "Writing cluster state: {}".format(workers))
+                    f.write(json.dumps(workers))
 
     def get(self):
-        with self.file_lock:
-            workers = json.loads(open(self.save_path).read())
-            return workers
+        with self.lock:
+            with self.file_lock:
+                workers = json.loads(open(self.save_path).read())
+                return workers
 
     def put(self, worker_id, info):
         assert "tags" in info
         assert "state" in info
-        with self.file_lock:
-            workers = self.get()
-            workers[worker_id] = info
-            with open(self.save_path, "w") as f:
-                logger.info("ClusterState: "
-                            "Writing cluster state: {}".format(workers))
-                f.write(json.dumps(workers))
+        with self.lock:
```

Comment:
i would just chain these in the same with statement, see: https://docs.python.org/2.7/reference/compound_stmts.html#with

Reply:
it would make the diff better and also reduce the indentations